### PR TITLE
Adds hsm declust heuristic and tests

### PR DESCRIPTION
**Pull request by nshea3**
Hi @juliohm! 

Copied the files to simplify the commit history here. Replaces PR #94 and closes [JuliaEarth/GeoStats.jl#177](https://github.com/JuliaEarth/GeoStats.jl/issues/177). 

Also adds the tests with the Gaussian mixtures that were missing in the other PR. 

Let me know if we'll need to make any adjustments! 

Thanks,

Nick

### CODE REVIEW

File: test/statistics.jl [hunk 1]
```diff
@@ -26,6 +26,19 @@
...
   end
 
+  @testset "HalfSampleMode" begin
+    d = LogNormal(0,1)
+    rs = rand(d,1000)
+    @test GeoStatsBase.hsm_mode([1,2,2,3]) == 2.0
+    @test GeoStatsBase.hsm_mode([1,2,2,3,5]) == 2.0
+    @test GeoStatsBase.hsm_mode(rs) < mean(rs)
+    @test GeoStatsBase.hsm_mode(rs) < median(rs)
+    d = Normal(1,0.5)
```
Comment (juliohm) [inline test/statistics.jl:36]: This `d` is not used?

File: test/statistics.jl [hunk 2]
```diff
@@ -26,6 +26,19 @@
...
+    rs = vcat(rand(Normal(0,1), 700), rand(Normal(3,0.2), 300))
```
Comment (juliohm) [inline test/statistics.jl:37]: Should we rewrite this as `d = MixtureModel([Normal(), Normal(3, 0.2)], [0.7, 0.3])` and then `rand(d, 1000)`?

**Comment by juliohm** [comment]
The tests are failing in Julia v1.5 probably because of some change in Plots.jl. I will remove the version of the language from the build bots, it is not very useful to have it here...

**Comment by nshea3** [comment]
Thanks for the suggestions, those are both incorporated in the latest version!

File: src/statistics/data.jl [hunk 3]
```diff
@@ -48,21 +48,74 @@ quantile(d, p::T, s::Number) where {T<:Union{Number,AbstractVector}} = quantile(
...
+"""
+    mode_heuristic(d)
+
+Return the estimated mode of the pairwise distances for a set of locations.
+"""
+function mode_heuristic(d)
+  D = dist_matrix_random_sample(d)
+  n = size(D, 1)
+  δ = [D[i,j] for i in 1:n for j in 1:n if i > j]
+  m = hsm_mode(δ)
```
Comment (juliohm) [inline src/statistics/data.jl:83]:
```suggestion
  m = hsm_mode([D[i,j] for i in 1:n for j in 1:n if i > j])
```

**Comment by juliohm** [comment]
I think it is time to merge this! 🎉 I will perform more tests locally with the GeoStatsTutorials on declustering to reproduce those results with the mode heuristic and we can try to setup a plan for comparing the heuristics with real data sets. What do you think? Should we exchange e-mail to further work on this for a specific venue next year? My e-mail is on my GitHub profile.

**Pull request merged**